File: frontend/src/types/market.ts
export type Quote = {
  symbol: string;
  price: number;
  change: number | null;
  change_percent: number | null;
  currency: string;
  market_cap: number | null;
  volume: number | null;
  previous_close: number | null;
  provider: string;
  freshness: string;
  as_of: string;
};

export type OHLCVBar = {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
};

export type OHLCVResponse = {
  symbol: string;
  interval: string;
  bars: OHLCVBar[];
  provider: string;
  freshness: string;
};

export type TechnicalIndicators = {
  symbol: string;
  rsi_14: number | null;
  macd: number | null;
  macd_signal: number | null;
  macd_histogram: number | null;
  ema_12: number | null;
  ema_26: number | null;
  sma_20: number | null;
  sma_50: number | null;
  sma_200: number | null;
  atr_14: number | null;
  vwap: number | null;
  bb_upper: number | null;
  bb_middle: number | null;
  bb_lower: number | null;
  adx_14: number | null;
  provider: string;
  computed_at: string;
  data_points: number;
};

export type Fundamentals = {
  symbol: string;
  pe_ratio: number | null;
  forward_pe: number | null;
  peg_ratio: number | null;
  eps: number | null;
  dividend_yield: number | null;
  beta: number | null;
  profit_margin: number | null;
  revenue_growth: number | null;
  sector: string | null;
  industry: string | null;
  provider: string;
  as_of: string;
};

export type AnalysisExplanation = {
  overall_rating: string;
  investment_thesis: string;
  reasons: string[];
  potential_risks: string[];
  confidence: number;
};

export type SymbolAnalysis = {
  symbol: string;
  quote: Quote | null;
  fundamentals: Fundamentals | null;
  technical: TechnicalIndicators | null;
  scores: {
    technical_score: number | null;
    momentum_score: number | null;
    risk_score: number | null;
  } | null;
  explanation: AnalysisExplanation | null;
  data_warnings: string[];
  disclaimer: string;
};

File: backend/app/schemas/market.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field

from app.schemas.fundamentals import AnalysisExplanation, Fundamentals


class DataFreshness(str, Enum):
    LIVE = "live"
    CACHED = "cached"
    DELAYED = "delayed"


class DataSourceStatus(str, Enum):
    OK = "ok"
    RATE_LIMITED = "rate_limited"
    MISSING_KEY = "missing_key"
    UNAVAILABLE = "unavailable"


class ProviderHealth(BaseModel):
    name: str
    status: DataSourceStatus
    message: str | None = None


class Quote(BaseModel):
    symbol: str
    price: float
    change: float | None = None
    change_percent: float | None = None
    currency: str = "USD"
    market_cap: float | None = None
    volume: int | None = None
    previous_close: float | None = None
    provider: str
    freshness: DataFreshness
    as_of: datetime


class OHLCVBar(BaseModel):
    timestamp: datetime
    open: float
    high: float
    low: float
    close: float
    volume: int


class OHLCVResponse(BaseModel):
    symbol: str
    interval: str
    bars: list[OHLCVBar]
    provider: str
    freshness: DataFreshness


class TechnicalIndicators(BaseModel):
    symbol: str
    rsi_14: float | None = None
    macd: float | None = None
    macd_signal: float | None = None
    macd_histogram: float | None = None
    ema_12: float | None = None
    ema_26: float | None = None
    sma_20: float | None = None
    sma_50: float | None = None
    sma_200: float | None = None
    atr_14: float | None = None
    vwap: float | None = None
    bb_upper: float | None = None
    bb_middle: float | None = None
    bb_lower: float | None = None
    adx_14: float | None = None
    provider: str
    computed_at: datetime
    data_points: int = Field(description="Number of OHLCV bars used in calculation")


class AnalysisScores(BaseModel):
    technical_score: float | None = Field(None, ge=0, le=100)
    momentum_score: float | None = Field(None, ge=0, le=100)
    risk_score: float | None = Field(None, ge=0, le=100)


class SymbolAnalysis(BaseModel):
    symbol: str
    quote: Quote | None = None
    fundamentals: Fundamentals | None = None
    technical: TechnicalIndicators | None = None
    scores: AnalysisScores | None = None
    explanation: AnalysisExplanation | None = None
    data_warnings: list[str] = Field(default_factory=list)
    disclaimer: str = (
        "Probabilistic research only — not financial advice. "
        "Figures sourced from live APIs; verify before trading."
    )


class HealthResponse(BaseModel):
    status: str
    app: str
    environment: str
    providers: list[ProviderHealth]
